update: protocols overview (#287)
libp2p/docs: content/concepts/introduction/protocols/overview.md

which negotiates the use of the protocol during the establishment
of a new stream between two peers.

Protocol IDs typically follow a path-like structure with a version number
as the final component.

```shell
/my-app/amazing-protocol/1.0.1
```

A new stream passes through a protocol multiplexer called
[Multistream-select](multistream.md), which routes the stream to the appropriate
protocol handler based on the protocol ID.

### Handler functions

A handler function handles each libp2p protocol, responsible
for defining the protocol's behavior once the stream has been established.
The handler function is invoked when an incoming stream is received with a
registered protocol ID.

The handler function can also specify a match function,
which allows for the acceptance of non-exact string matches for protocol IDs.

A libp2p application will define a stream handler that takes over the
stream after protocol negotiation. Everything is sent and received after the
application protocol defines the negotiation phase.

### Binary streams

The medium over which a protocol operates is a bi-directional
binary stream. This stream provides the following features:

> - **Bidirectional, reliable delivery of binary data**: Both peers can read and write

Reviewer: ```suggestion
The primary benefit provided to libp2p protocols by the stream abstraction is:
**Bidirectionality and reliable delivery of binary data**: Both peers can read and write
```

  from the stream at any time, and data is read in the same order as it was written.
  The stream can also be "half-closed", meaning it can be closed for writing while
  still open for reading or closed for reading while still open for writing.

> - **Supports backpressure**: Eager writers cannot flood readers with data, as the
>   stream automatically regulates data flow to prevent overload.

Reviewer: ```suggestion
- **Supports backpressure**: Eager writers cannot flood readers with data, as the
  stream automatically regulates data flow to prevent overload.
```
Not by default, if TCP + mplex is used, there's no support for backpressure. I think this existed in the legacy document but it should be deleted

> Behind the scenes, libp2p also ensures that the stream is securely encrypted and
> efficiently multiplexed, allowing multiple logical streams to be multiplexed over

Reviewer: ```suggestion
Behind the scenes, libp2p also ensures that the stream is securely encrypted and
efficiently multiplexed, allowing multiple logical streams to be multiplexed over
```
link to secure channel and muxer docs?

a single underlying connection. These details are transparent to the protocol handler,
which reads and writes unencrypted binary data over the stream.

> ## Life cycle of a stream
> 
> The life cycle of a libp2p stream involves the following steps:
> 
> - **Dialing out**: When a peer wants to initiate a new stream with another peer,
>   it sends the protocol ID of the protocol it wants to use over the connection.
> - **Protocol negotiation**: The listening peer on the other end checks the incoming
>   protocol ID against its list of registered protocol handlers. Suppose it does not
>   support the requested protocol. In that case, it sends "na" (not available) on the stream.
>   The dialing peer can try again with a different protocol or a fallback
>   version of the initially requested protocol. If the protocol is supported, the
>   listening peer echoes the protocol ID as a signal that future data sent over
>   the stream will use the agreed-upon protocol semantics.
> - **Stream establishment**: Once peers agree on the protocol ID, the stream is
>   established, and the designated invokes the handler function to take over the stream.
>   Everything sent and received over the stream from this point on is defined by the
>   application-level protocol.
> - **Stream closure**: When either peer finishes using the stream, it can be closed
>   by either side. If the stream is half-closed, the other side can continue to read
>   or write until it is closed.

Reviewer: ```suggestion
```
I would move this to the stream multiplexing document in a stream overview subsection
Is the protocols overview a good place for this?

---
title: "Overview"

> description: There are protocols everywhere you look when you're writing network applications, and libp2p is especially thick with them.

Reviewer: I like the zest of this earlier description, can we keep it?

Author: yes

> title: "What is a libp2p Protocol"

Reviewer: ```suggestion
title: "What is a libp2p Protocol?"
```
should this have a `?`

Author: nope

Author: We usually try to leave out the quotation marks in the sidebar as it is cleaner, but they are present with in other headings and subheadings.

However, I've been meaning to get around to this to add logic for us to define the sidebar title separate from the title (like in other generators). I don't believe you can do this in Hugo by default.

weight: 20
aliases:
    - "/concepts/protocols"
    - "/concepts/fundamentals/protocols"
---

libp2p is composed of various core abstractions, such as
[peer identity](../core-abstractions/peers.md#peer-id),
and [addressing](../core-abstractions/addressing.md/), and
relies on protocols to facilitate communication between peers.

A libp2p protocol is a network protocol that operates within

> the libp2p network and follows certain conventions.

Reviewer: I think the previous intro was better and did a good job, the new lines are a bit too succinct imo.

> A libp2p protocol is a network protocol that operates within the libp2p network and follows certain conventions.

This is a very circular definition. 😵‍💫

Author: That's fair.

This is trying to be direct and point out *exactly* what a libp2p protocol is. I simplified it.
